### PR TITLE
PR: Several Find in Files improvements

### LINKED ISSUE #1555: Find in Files unusable
_From [peter.sc...@synerscope.com](https://code.google.com/u/105847980652508035257/) on 2013-09-11T04:17:48Z_

Spyder Version:  2.1.13.1
Python Version:  2.7.4
Qt Version:      4.8.4, PyQt4 (API v2) 4.10 on Linux 

What steps will reproduce the problem?
1. Ctrl-Shift-F 2. 3. What is the expected output? What do you see instead? Expect to see a useful, understandable dialog.
   Instead, I don't know what this is supposed to be. 

Please provide any additional information below

. The search-in-files, I'm sorry to say, is a f*cking disgrace!
- Errors are not cleared at new search, so I don't know if it's busy or the same error occurs
- It can't search in open files only
- When I hide advanced options, it secretly 
- It searches a complete directory structure, useless when things like 3-rd party tool dirs are present
- Options and buttons are all over the place
- No feedback on what it's doing.
- No hints on what to put in the search fields
- No research on ENTER
- I somehow entered wrong character, but can't for the love of god get the search to work again.
- etc.

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=1555_

### LINKED ISSUE #2730: Find in files default include and exclude values should be empty!
The find in files default exclude and include values should be empty? And offer a better way that having a predetermined list of defaults, like a pop up table or something more friendly for a GUI

include = .pyw?$|.ipy$|.txt$|.rst$
exclude = .pyc$|.pyo$|.orig$|.hg|.svn|\bbuild\b

image

I was bitten by this by not finding stuff in a file named `build.py` due to the nature of the exclude list. I would take the build out at least.

@spyder-ide/core-developers ideas?

### LINKED ISSUE #2850: Adapt the Find in Files so results display is flat, like ACK
# Example screenshot

image
# Todo
- [ ] Make tree a flat structure
- [ ] Add toggling options
- [ ] Highlight matched words

### LINKED ISSUE #2963: Find in files results disappear when search path loses focus
When doing a search in the "Find in files" panel, the panel seems to have a bug in the signal handling.

When I open the panel with no item in the search path combo box, the field is pre-filled with the current working directory or the previously searched path. But as soon as I click on the search field to type what I want to search, the search path field is emptied. So even if I was fine with the pre-filled value I have to re-select it. This is the first problem (minor).

Then when I select one item from the combo box (or type one), the search is automatically triggered and results are displayed. If I then try to click on the results (or anywhere else actually) **without validating the entry in the search path field** (with Enter key), the search path field is emptied, and the search results refreshed, that is: emptied also. This is the second problem.
It took me a while to figure out that I have to press Enter to be able to read the search results.

This happens in latest Spyder 3.0.0dev, from what I remember it was working fine in Spyder 2.

I tried to have a look at the code in `findinfiles.py`, but got a bit lost in all the signals emitted by the different components. From what I could understand, I think that when the search path entry is not validated and loses focus, a signal is emitted that will cause the field to be reset with a default value of None and eventually re-run the search. Since the path is now empty, the search has no result and the results box is emptied. It is likely I think that the two problems are related.
I can try to have a better look if someone can give me some pointers or explanation of who emits what signal and when.

### LINKED ISSUE #4134: Non-ascii characters not displayed correctly in Find in files
## Description

**What steps will reproduce the problem?**

1. Write a line containing a non-ascii character in a file, e.g. `str = 'Viva España'`
2. Use 'Find in files' to find that line, e.g. search for `Viva`

**What is the expected output? What do you see instead?**

I would have expected the non-ascii character to be displayed correctly in the "Find in files" pane. Instead I see a hex representation.

The below screenshot hopefully makes it clear. I would have liked to see `España` in the location marked with the red circle.

Screenshot

## Version and main components

* Spyder Version: 3.2.0.dev0 847ea1a2
* Python Version: 3.6.0
* Qt Versions: 5.6.2, PyQt5 5.6 on Linux

## Dependencies
```
pyflakes >=0.6.0 :  1.5.0 (OK)
pep8 >=0.6       :  1.7.0 (OK)
pygments >=2.0   :  2.1.3 (OK)
qtconsole >=4.2.0:  4.2.1 (OK)
nbconvert >=4.0  :  4.2.0 (OK)
pandas >=0.13.1  :  0.19.2 (OK)
numpy >=1.7      :  1.11.3 (OK)
sphinx >=0.6.6   :  1.5.1 (OK)
rope >=0.9.4     :  0.9.4-1 (OK)
jedi =0.9.0      :  0.9.0 (OK)
psutil >=0.3     :  5.0.1 (OK)
matplotlib >=1.0 :  2.0.0 (OK)
sympy >=0.7.3    :  1.0 (OK)
pylint >=0.25    :  1.6.4 (OK)
```

### LINKED ISSUE #4197: Find in Files not working in 3.1.3
## Description
Should have been fixed by PR #4144  
Same as issue #4005 
Issue persists

## Version and main components

* Spyder Version: 3.1.3 
* Python Version: 3.5.2
* Qt Versions: 5.6.0, PyQt5 5.6 on Linux

## Dependencies
```
pyflakes >=0.6.0 :  1.3.0 (OK)
pep8 >=0.6       :  1.7.0 (OK)
pygments >=2.0   :  2.1.3 (OK)
qtconsole >=4.2.0:  4.2.1 (OK)
nbconvert >=4.0  :  4.2.0 (OK)
pandas >=0.13.1  :  0.18.1 (OK)
numpy >=1.7      :  1.11.1 (OK)
sphinx >=0.6.6   :  1.4.6 (OK)
rope >=0.9.4     :  0.9.4-1 (OK)
jedi =0.9.0      :  0.9.0 (OK)
psutil >=0.3     :  5.0.0 (OK)
matplotlib >=1.0 :  1.5.3 (OK)
sympy >=0.7.3    :  1.0 (OK)
pylint >=0.25    :  1.6.4 (OK)

```

### PR DESCRIPTION
**Pull request by andfoy**
Fixes #4197
Fixes #4134
Fixes #2963
Fixes #2850
Fixes #2730
Fixes #1555   

----

After checking issue #4005, I've identified that the widget in it's current incarnation doesn't allow to track the status of the search, that implies that the User doesn't know if the search is taking place or not, especially when the search is launched in a big directory. Taking in account this requirement, a progress bar and a text field that indicates the current file scanned is added. Also, some efficiency issues and codification mismatches are being tackled in the present PR.

- [x] Add spinner and indicator of search progress 
- [x] Inform user if the search/exclude/include regexp patterns are incorrect
- [x] Improve search and scanning of files (Do not recover all files before searching a term)
- [x] Improve file and string encoding to bypass and correct errors associated to codification
- [x] Fix Spyder crash after a long line is matched and tried to display on the results window
- [x] Simplify visualization of results (The current tree directory visualization is useless if a file resides inside a deep folder)
- [x] Break results if the pattern was found in a single line 
- [x] Display results as search takes place
- [x] Replace default regex suggestions
- [x] Remove previous search results when a new search takes place
- [x] Remove unused search options
- [x] Search path now refers to global explorer path
- [x] Omit binary files
- [x] Add options to search on the current file, project or current path
- [x] Add test
- [x] Allow to order results alphabetically from top header

Now the Widget should look like this while a search task is taking place
peek 25-02-2017 16-25

### CODE REVIEW

**Comment by ccordoba12** [comment]
@andfoy, the progress bar looks great!

However, you need to break this PR in two parts: one that goes against `3.1.x` and fixes only issue #4005 and another one against `3.x` with the new additions.

**Comment by ccordoba12** [comment]
You can leave this one as it is, then rebase against `3.x` once we merge the fix for #4005.

**Comment by andfoy** [comment]
@ccordoba12 This long PR is ready!

**Comment by andfoy** [comment]
@ccordoba12 I have some errors on CircleCI unrelated to my tests

**Comment by ccordoba12** [comment]
I restarted CircleCI to see if that fixes the errors.

**Comment by andfoy** [comment]
This error persists: 
```
spyder/plugins/tests/test_ipythonconsole.py::test_load_kernel_file_from_location ./continuous_integration/circle/test-qt5.sh: line 13: 13574 Segmentation fault      (core dumped) python runtests.py
```

**Comment by andfoy** [comment]
@ccordoba12 Finally, I think this PR is ready to go

**Comment by goanpeca** [comment]
@andfoy can you rebase or merge with latest 3.x to solve conflicts?

File: spyder/plugins/findinfiles.py [hunk 1]
```diff
@@ -112,7 +124,8 @@ def include_patterns():
                     r'\.ipy$|\.pyw?$|\.rst$|\.txt$',
                     '.',
                     ]
-        return patterns
+        # return patterns
+        return []
```
Comment (ccordoba12) [inline spyder/plugins/findinfiles.py]: Why this returns an empty list? Isn't it better to simply remove this method?

File: spyder/widgets/findinfiles.py [hunk 2]
```diff
@@ -13,102 +13,43 @@
 
 # Standard library imports
 from __future__ import with_statement
+from __future__ import unicode_literals
```
Comment (ccordoba12) [inline spyder/widgets/findinfiles.py]: `unicode_literals` introduces a lot of problems. Please remove it.

Comment (ccordoba12) [inline spyder/widgets/findinfiles.py]: Also, you can simplify importing from `__future__` like this

    from __future__ import with_statement, print_function

File: spyder/widgets/findinfiles.py [hunk 3]
```diff
@@ -126,90 +67,42 @@ def __init__(self, parent):
         self.text_re = None
         self.completed = None
         self.get_pythonpath_callback = None
-        
-    def initialize(self, path, python_path, hg_manifest,
-                   include, exclude, texts, text_re):
+        self.results = {}
+        self.nb = 0
```
Comment (ccordoba12) [inline spyder/widgets/findinfiles.py]: Please use a more descriptive name instead of `nb`.

File: spyder/widgets/findinfiles.py [hunk 4]
```diff
@@ -533,161 +485,125 @@ def __init__(self, parent):
...
-                    text_files += 's'
-                text = "%d %s %d %s" % (self.nb, text_matches,
-                                        nb_files, text_files)
-        if self.error_flag:
-            text += ' (' + self.error_flag + ')'
-        elif self.results is not None and not self.completed:
-            text += ' (' + _('interrupted') + ')'
-        self.set_title(title+text)
+
+    def init(self, search_text):
```
Comment (ccordoba12) [inline spyder/widgets/findinfiles.py]: This is a bad name, please use a better one (I mean, there's already `__init__` and `initialize`)

**Comment by ccordoba12** [comment]
@andfoy, I left some minor comments. After you solve them, I think this is ready to merge!

But before doing any changes, please remove your last 3 commits (the ones that merge with `3.x`) and merge again with the latest 3.x (but just one time).

**Comment by ccordoba12** [comment]
@andfoy, the last two commits you added are a mess! I'm going to remove them, and do a proper merge with 3.x.

After that you could fix my comments from yesterday.

**Comment by andfoy** [comment]
@ccordoba12 I have some CI errors

**Comment by ccordoba12** [comment]
You're trying to merge your changes from yesterday with my changes from today, which is again introducing the mess I mentioned before.

Please revert what you did (i.e. your last merge) and **redo** again your changes on top of my last commit on this branch (i.e. `Testing: Fix tests`). Don't do rebases or merges, just make your changes again (by hand, sorry for making you waste time, but that's what's needed here).

**Comment by andfoy** [comment]
The error persists, even if the conflicting merge was removed

**Comment by ccordoba12** [comment]
Ok, don't worry. I'm seeing the same error locally.

I'll find a fix, push it and merge your PR. Thanks a lot for your patience :-)

**Comment by ccordoba12** [comment]
This long PR is (finally!) ready. Thanks @andfoy for your work!

**Pull request merged**